BCP00301Test.py

```diff
@@ -117,3 +120,140 @@ def test_02(self, test):
                                     .format(",".join(tls1_3_should)))
             else:
                 return test.PASS()
+
+    def test_03_cn_san(self, test):
+        """Certificate does not use IP addresses in CN/SANs"""
+
+        ret = self.perform_test_ssl(test, ["-S"])
+        # Known issue: If the OCSP URL is unreachable testssl generates invalid JSON and exits with a non-zero code
+        if ret != 0:
+            return test.FAIL("Unable to test. See the console for further information.")
+        else:
+            common_name = None
+            with open(TMPFILE) as tls_data:
+                tls_data = json.load(tls_data)
+                for report in tls_data:
+                    if report["id"].split()[0] == "cert_commonName":
+                        common_name = report["finding"]
+                        try:
+                            ipaddress.ip_address(report["finding"])
+                            return test.WARNING("CN is an IP address: {}".format(report["finding"]))
+                        except ValueError:
+                            pass
+                    elif report["id"].split()[0] == "cert_subjectAltName":
+                        if report["finding"].startswith("No SAN"):
+                            return test.WARNING("No SAN was found in the certificate")
+                        else:
+                            alt_names = report["finding"].split()
+                            if common_name not in alt_names:
+                                return test.FAIL("CN {} was not found in the SANs".format(common_name))

```

Comment:
In the spec this is a SHOULD, not a SHALL, so maybe this should be a warning?

Reply:
Good point. I'll fix that.

Of note in general, I've used 'warning' rather than 'optional' in these cases. Don't know if you have any opinion on this @garethsb-sony. I guess in some ways I'm trying to differentiate a little between SHOULD and MAY.

Reply:
There's definitely some overlap in our description of `WARNING` and `OPTIONAL`. If I understand correctly, in this particular case it definitely looks like a misconfiguration which I agree should be a `WARNING`, rather than an _unimplemented_ optional feature? Would focusing on that differentiate between which should be used generally?

For both `WARNING` and `OPTIONAL`, it would be good to put in the links to the spec or wiki as we go.

I also wonder if there are a few `FAIL`s that could be indicated as `DISABLED`, if the failure might well be nothing to do with API under test. E.g. if testssl.sh just fails to return a response or one that can't be parsed?

Reply:
Ok, I'll have a scan through what's there on that basis and see what I can do.